### ISSUE TITLE
PSCmdlet.ThrowTerminatingError() does not result in a terminating error

### ISSUE THREAD

**Issue by colinblaise** [PowerShell/PowerShell#19599]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Run the attached repro project.

[powershelltest.zip](https://github.com/PowerShell/PowerShell/files/11374696/powershelltest.zip)

### Expected behavior

```console
An exception should occur in the pipeline as a terminating error
```

### Actual behavior

```console
A terminating error is not thrown
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.3.4
PSEdition                      Core
GitCommitId                    7.3.4
OS                             Microsoft Windows 10.0.22000
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by colinblaise** [comment]
Thanks for the response.  Yeah,  I understand it's just how it was designed.  It's just bizarre that "throw" when called directly in PowerShell doesnot take into account e.a.p, but "throw" in a C# PSCmdlet does.

**Comment by jhoneill** [comment]
We'll get into angels on the head of pin, but ... `Throw` produces errors which don't consider themselves handled by printing the message (which is the default for e.a.p) but if an `$errorActionPreference` setting of `silentlyContinue` inherits into (or is explicitly set for) a function containing `throw` the `throw` is considered to be handled there and then - which can have unexpected consequences. throw was designed for one script to throw the anchors out and it's caller, and it's caller's caller etc.

_ThrowTerminatingError_ was designed for a (compiled) cmdlet to say "er, we got a problem, we're bailing, you might want to think about aborting" - the caller's e.a.p decides what happens (so `-errorAction` doesn't change anything) and the default action preference is "yeah there was an alarm, keep going".

**Comment by SeeminglyScience** [comment]
`ThrowTerminatingError` generates a *pipeline* terminating error rather than a call stack terminating error. Only the pipeline that the command is in is terminated, subsequent statements will continue. `ErrorAction` does not have any effect on pipeline terminating errors. Also a binary cmdlet cannot generate call stack terminating errors, that's not supposed to be the job of a command.

**Comment by mklement0** [comment]
@colinblaise, _unofficial_ terms have been coined - which are hopefully descriptive enough, even though they're not fully technically correct - to help conceptualize the behavior:

_In effect_, PowerShell has _two_ types of terminating behavior, as implied by @SeeminglyScience's comment:

* _Statement_-terminating errors (pipeline-terminating errors), which - by default -  _continue_ execution (at the _next_ statement/pipeline).

* _Script_-terminating (call-stack-terminating, runspace-terminating), which - by default - are _fatal_.

Unfortunately, this crucial distinction is still missing from the documentation.

I've attempted to summarize the inconsistencies and complexities of PowerShell's error handling in the following issue, which has become a live document of sorts (do tell us if something doesn't look right):

* https://github.com/MicrosoftDocs/PowerShell-Docs/issues/1583

**Comment by jhoneill** [comment]
@mklement0  I know you have used "Statement terminating" for some time , and  it is sometimes difficult  language between a language statement (for, while etc.) a command (alias, function, cmdlet, external program) and a pipeline (command or value(s) or a pipeline [but not for or similar statements], optionally piped into a command) 

The code below shows that ThrowTerminatingError stops commands earlier in the pipeline, i.e. it is pipeline terminating, but it does not stop a foreach _statement_

```PowerShell
function foo {
    [CmdletBinding()]
    param (
        [Parameter(ValueFromPipeline=$true,Position=0)]
        $p
    )
    process {
        if ($p -gt 5) {
            $er = [System.Management.Automation.ErrorRecord]::new("Invalid P","99",[System.Management.Automation.ErrorCategory]::InvalidData, $null )
            $pscmdlet.ThrowTerminatingError($er)
        }
        else     { $p }
    }
}


function bar {
    [CmdletBinding()]
    param (
        [Parameter(ValueFromPipeline=$true,Position=0)]
        $p
    )
    process {
    Write-host "$p in bar"
     $p 
    }
}


 1..7 | bar | foo | foreach {$_ * $_}
 

foreach ($v in (1..7)) {foo $v}
```

It might be better to stick to "Pipeline terminating" for these errors.

**Comment by mklement0** [comment]
**tl;dr**

* As long as an official name is chosen and documented, I'd be fine with _pipeline_-terminating, though my concern is that end users don't think of expressions such as `1 / 0` or `$x = [int]::Parse('foo')` as _pipelines_, and, despite the potential confusion with _language statement_, there is precedent for naming the smallest unit of execution a _statement_.

* Plus, if such an error occurs in the _conditional_ part of a language statement, it _does_ terminate the entire construct; e.g.:

```powershell
# The inputs are evaluated up front, and the statement-terminating
# error terminates the entire compound statement, without entering the loop body.
foreach ($i in 1, (1/0), 3) {
    $i
}
```

---

@jhoneill, let me start with what I think is the fundamental problem:

* Fundamental concepts are missing from the documentation.
* When they are touched upon, they aren't discussed as such _and aren't given official names_.

Things have improved over the years (think _delay-bind script parameters_, _calculated properties_ and, most recently, _member-access enumeration_), though in the realm of error handling the documentation is still woefully lacking.

Once _official_ names have been established and concepts are documented using those names, how intuitive those names are isn't all that important anymore - although certainly preferable.

---

As for _statement_-terminating vs. _pipeline_-terminating:

I originally picked _statement_-terminating based on [this homegrown taxonomy](https://github.com/MicrosoftDocs/PowerShell-Docs/issues/1583#issuecomment-368673114) of PowerShell execution units.

While the term _language statement_ for constructs such as `foreach`, `while`, ... is _somewhat_ established:

* it _isn't_ used in the conceptual [about_Language_Keywords](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_Language_Keywords) help topic, where the term _statement_ in isolation is used to refer to a statement in the sense that I had in mind: as the smallest unit of execution, as evidenced by referring to the execution units that such a language statement _is composed of_; e.g, the sample code for `continue` shows a `while` command composed of `<statements>` parts.  
And while you can make the argument that this is just a recursive definition, this just shifts the goal post: ultimately, there must be unit of execution other than a language statement that they are composed of.

* by contrast, topics such as [about_While](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_While) do talk about _language statements_ but also refer to their constituent parts as _statements_.

In short: the distinction is fuzzy, but my _guess_ is that the term _statement_ in isolation - without the _language_ qualifier - will be understood to mean _the smallest unit of execution_, be it a single expression or a single command / command pipeline (possibly in combination with an assignment) - as opposed to a whole "group" of such statements, as enclosed in a keyword-based flow-control construct such as `while` (I've used the word _compound statement_ in the past, which strikes me as conceptually clearer, but it seems that _language statement_ has won out.)

**Comment by jhoneill** [comment]
> @jhoneill, let me start with what I think is the fundamental problem:
> 
> * Fundamental concepts are missing from the documentation.
> * When they are touched upon, they aren't discussed as such _and aren't given official names_.

We are in complete agreement here.
